File: 0519/15686_치킨배달_백준.js
const readline = require('readline')
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
})

input = []

solution = () => {
    const temp = input[0].split(" ").map((el) => parseInt(el))
    n = temp[0]
    m = temp[1]
    
    let city = []
    for (let i = 1; i < n + 1; i++) {
        city.push(input[i].split(' ').map(Number))
    }
    house = []
    len_h = 0
    chicken = []
    let len_c = 0
    for (let row = 0; row < n; row++) {
        for (let col = 0; col < n; col++) {
            if (city[row][col] === 1) {
                house.push([row, col])
                len_h++                
            } else if (city[row][col] === 2) {
                chicken.push([row, col])
                len_c++
            }
        }
    }
    result = Infinity
    my_com(0, len_c, m, [])
    console.log(result)
}

my_com = (k, n, r, com) => {
    if (com.length === r) {
        let total = 0
        let isOK = true
        for (let i = 0; i < len_h; i++) {
            total += shortCut(house[i], com)
            if (result < total) {
                isOK = false
                break
            }
        }
        if (isOK) {
            result = total
        }
        return;
    }
    if (k === n) return;

    my_com(k + 1, n, r, com)
    com.push(chicken[k])
    my_com(k + 1, n, r, com)
    com.pop()
    
}

distance = (n1, n2) => {
    let r1 = n1[0]
    let r2 = n2[0]
    let c1 = n1[1]
    let c2 = n2[1]
    return Math.abs(r2 - r1) + Math.abs(c2 - c1)
}

shortCut = (n, c) => {
    let short = Infinity
    c.forEach(element => {
        let d = distance(element, n)
        if (short > d) {
            short = d
        }
    })
    return short
}

rl.on('line', line => input.push(line))
.on('close', () => {
    solution()
    process.exit()
})

File: 5월/0519/15686_치킨배달_백준.py
# 1. 가장 가까운 거리에 있는 치킨집 거리가 치킨거리다.
# 2. 하지만 치킨집 중에 M개만 살려야 한다.
# 3. 따라 조합으로 치킨집을 뽑아서 가장 가까운 거리를 구하는게 어떨까?
# 4. 야비하게 이터툴즈를 써서 문제를 풀어보자

import sys
sys.stdin = open('15686.txt')
from itertools import combinations

def distance(n1, n2):
    r1, c1 = n1
    r2, c2 = n2
    return abs(r2 - r1) + abs(c2 - c1)

def shortcut(n, c):
    short = float('inf')
    for i in c:
        d = distance(n, i)
        if short > d:
            short = d
    return short

N, M = map(int, input().split())

city = [list(map(int, input().split())) for _ in range(N)]

house = []
len_h = 0
chicken = []
len_c = 0

for row in range(N):
    for col in range(N):
        if city[row][col] == 1:
            house.append((row, col))
            len_h += 1
        elif city[row][col] == 2:
            chicken.append((row, col))
            len_c +=1

result = float('inf')

for combi in combinations(chicken, M):
    total = 0
    for h in house:
        total += shortcut(h, combi)
        if result < total:
            break
    else:
        result = total

print(result)